---
title: "Deconvolution Methylation data - MeDeCom"
author: "Sarah Hatzenbuheler"
date: "26 11 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Packages

```{r}
library(devtools)
#devtools::install_github("lutsik/MeDeCom",ref="windows")
library(MeDeCom)
```

# Example data set and how to proceed

```{r}
#load in example data
data(example.dataset, package="MeDeCom")
```
```{r}
# load beta matrix from preprocessed data
setwd("C:/Users/Sarah/Documents/Studium/Master/Systems Biology Maastricht/Thesis/R")
beta_mat <- as.matrix(read.table(file = "beta_preprocessed_teo.txt"))
```

 # Running Deconvolution
 
Before running the deconvolution, it is important to think about a number of parameters.
Ks : number range of cell types included in the mix
lambdas : range of regularisation parameter, with 0 there being no regularisation. Lambda basically restricts any possible beta values to be between 0 and 1, thus confirming closer to biology.
NINIT : number of random initialisations
NFOLDS : number of crossvalisation folds
NCORES: number of CPU cores the decon is running on
(for example numbers check the vignette: https://github.com/lutsik/MeDeCom/blob/master/vignettes/MeDeCom.md)

According to Weronicka's thesis: NINIt = 30-50

```{r}
# Running medecom decon
#medecom.result<-runMeDeCom(D, Ks = 2:5, lambdas = c(0,10^(-2:-1)), 
#                           NINIT = 2, NFOLDS = 2, ITERMAX=300, NCORES=5)

medecom.result<-runMeDeCom(beta_mat, 2:10, c(0,10^(-5:-1)), NINIT=10, 
                           NFOLDS=10, ITERMAX=300, NCORES=5)
```

# Plot results

```{r}
# visaulisa CV error for different lambda
plotParameters(medecom.result)
```
To select optimal lambda for different k, a different visualisation can be employed:
```{r}
# line plot for set k
plotParameters(medecom.result, K=4, lambdaScale="log")
```
# Extractiong Latent Methylation components

```{r}
# extract LMCs for set lambda and K
lmcs<-getLMCs(medecom.result, K=4, lambda=0.01)
str(lmcs)
```

# Visualisation of LMCs
```{r}
# dendrogram
plotLMCs(medecom.result, K=4, lambda=0.01, type="dendrogram")

# MDS plot with data included
plotLMCs(medecom.result, K=4, lambda=0.01, type="MDS", D = beta_mat)
```
# matching LMCs to reference profiles
```{r}
# dendrogram
plotLMCs(medecom.result, K=5, lambda=0.01, type="dendrogram", Tref=Tref, center=TRUE)

# heatmap
plotLMCs(medecom.result, K=5, lambda=0.01, type="heatmap", Tref=Tref)
```

# Get mixing proportions matrix for RNAseq analysis

```{r}
# get matrix of mixing proportions
prop<-getProportions(medecom.result, K=4, lambda=0.01)
str(prop)
```

```{r}
#create stacked bar plots of proportions

prop<-getProportions(medecom.result, K=4, lambda=0.01)

# create color palette:
library(RColorBrewer)
coul <- brewer.pal(4, "Set1") 
 
# Transform this data in %
data_percentage <- apply(prop, 2, function(x){x*100/sum(x,na.rm=T)})
 
# Make a stacked barplot--> it will be in %!
barplot(data_percentage, col=coul , border="white", xlab="samples", ylab = "cell type %",
        main="cell type proportion for FS based on variance & Weronicka's gene list")
```


